File: java/2015/day08b/Program.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.regex.*;

public class Program {
    public static void usage() {
        System.err.println("usage: java -jar Program.jar <input file>");
        System.exit(1);
    }

    public static int process(String filename) {
        int result = 0;
        try ( BufferedReader reader = new BufferedReader(new FileReader(filename))) {
            String line;
            while ( (line = reader.readLine()) != null ) {
                int codeLen = line.length();
                int encLen = 0;
                int i = 0;
                for ( i = 0; i < codeLen; i++ ) {
                    switch ( line.charAt(i) ) {
                    case '\\':
                    case '\"':
                        encLen += 2;
                        break;
                    default:
                        encLen += 1;
                    }
                }
                result += 2 + (encLen - codeLen);
            }
        } catch ( IOException e ) {
            System.err.println("Error reading file: " + e.getMessage());
            System.exit(1);
        }
        return result;
    }

    public static final void main(String[] args) {
        if ( args.length < 1 ) {
            usage();
        }
        String filename = args[0];
        int result = process(filename);
        System.out.println("result = " + result);
    }
}

File: kotlin/2015/day08b/Program.kt
import java.io.File

fun usage() {
    System.err.println("usage: java -jar Program.jar <input file>")
    kotlin.system.exitProcess(1)
}

fun process(content: String) : Int {
    var result = 0

    for ( line in content.split("\n") ) {
        if ( line.length > 0 ) {
            val codeLen = line.length
            var encLen = 0
            for ( i in 0 until line.length ) {
                when ( line.get(i) ) {
                    '\\', '"' -> encLen += 2
                    else -> encLen += 1
                }
            }
            result += 2 + (encLen - codeLen)
        }
    }
    return result
}

fun main(args: Array<String>) {
    if ( args.isEmpty()) {
        usage()
    }
    val filename = args[0]
    try {
        val content = File(filename).readText()
        val result = process(content)
        println("result = $result")
    } catch ( e: Exception ) {
        System.err.println("error reading file: ${e.message}")
    }
}
